src/matrix_assembler.cc

```diff
@@ -31,111 +35,55 @@ void pipenetwork::MatrixAssembler::assemble_variable_vector() {
-  // equation. If any of the nodal head is unknown, initialize the
-  // discharge to 0.001
   for (const auto& pipe : global_pipes_) {
     Index index_pd = pipe.first + 2 * nnode_;
     variable_vec_->coeffRef(index_pd) = pipe.second->discharge();
   }
 }
 
-// Apply variables (head and discharge) back to nodes and pipes
-void pipenetwork::MatrixAssembler::apply_variables() {
-
-  // Iterate through nodes, assign nodal variables
-  for (auto& node : global_nodes_) {
-    // Get index of target variables
-    Index index_nh = node.first;
-    Index index_nd = node.first + nnode_;
-    // Assign head
-    node.second->head(variable_vec_->coeff(index_nh));
-    // Assign discharge
-    node.second->discharge(variable_vec_->coeff(index_nd));
-  }
-
-  // Iterate through pipes, assign pipe discharge during iteration
-  for (auto& pipe : global_pipes_) {
-    // Get index of target variables
-    Index index_pd = pipe.first + 2 * nnode_;
-    // Assign discharge
-    pipe.second->iter_discharge(variable_vec_->coeff(index_pd));
-  }
-}
-
-// Calculate and assemble residual vector
-// Nodal balance equation:
-// Residual = flow = Inflow - Outflow - Demand
-// Headloss equation (Hazen-Williams):
-// Residual = (Start_node_head - end_node_head) - \frac{10.67 \times length
-//            \times pipe_discharge^1.852}{pipe_roughness^1.852 \times
-//            (2radius)^4.8704}
-// The network has nnode Nodal balance equation and npipe Headloss equation
-// Thus the residual vector has (nnode+npipe) elements
-void pipenetwork::MatrixAssembler::assemble_residual_vector() {
-  residual_vec_->resize(nnode_ + npipe_);
-  residual_vec_->setZero();
-
-  // Iterate through all pipes
-  for (auto& pipe : global_pipes_) {
-    Index index_pipe = pipe.first;
-    Index index_node1 = pipe.second->nodes().at(0)->id();
-    Index index_node2 = pipe.second->nodes().at(1)->id();
-    // Calculate headloss residual
-    pipe.second->compute_headloss_hazen_williams();
-    double head1 = 0;
-    double head2 = 0;
-    if (pipe.second->nodes().at(0)->ishead())
-      head1 = pipe.second->nodes().at(0)->head();
-    if (pipe.second->nodes().at(1)->ishead())
-      head2 = pipe.second->nodes().at(1)->head();
-    residual_vec_->coeffRef(nnode_ + index_pipe) =
-        (-1) * ((head1 - head2) - pipe.second->headloss());
-    // Calculate the nodal balance residual (in and out flow part)
-    residual_vec_->coeffRef(index_node1) +=
-        (-1) * (-1) * pipe.second->iter_discharge();
-    residual_vec_->coeffRef(index_node2) +=
-        (-1) * pipe.second->iter_discharge();
-  }
-
-  // Iterate through all nodes
-  for (const auto& node : global_nodes_) {
-    Index index = node.first;
-    // Calculate the nodal balance residual (demand part)
-    if (node.second->isdischarge())
-      residual_vec_->coeffRef(index) -= (-1) * node.second->discharge();
+// assember the pressure demand part of jacobian matrix
+void pipenetwork::MatrixAssembler::construct_demand_jac(
```

Comment:
Can we not return a vector instead of passing it as a reference? if we are going to modify it?

Reply:
What you mean about returning a vector?

Reply:
instead of void could we return?

Reply:
I am trying to construct the sparse matrix. Is it easier to maintain a single triplets vector here? Or you mean return a vector then append this vector to original one? Whats the difference?